File: hw6/Driver.c
/****************************************************************************

                                                Tianling Luo
                                                CSE 12, winter 2020
                                                Feb 11, 2020
                                                cs12wi20cu
                              Assignment 6

File Name:      Driver.c
Description:    In this assignment, I will write the class implementation
for UCSDStudent which includes all UCSDStudent methods that will be 
called by the HashTable code.
****************************************************************************/

#include <cstdlib>  
#include <cstdio>  
#include <cstring>  
#include <getopt.h>  
#include <iostream>  
#include "Base.h"  
#include "SymTab.h"  
using namespace std;

#ifdef NULL  
#undef NULL  
#define NULL 0  
#endif  

/****************************************************************************
// class UCSDStudent
//
// Description: The UCSDStudent class is defined and it contains the member
// functions that will be called by the Hash.c code.
//
// Data Fields:
//     name (char*)  - the name of UCSD student
//     Studentnum (long)  - the student number of UCSD student
//
// Public functions:
//     UCSDStudent - store the name and student number in name and
//     		     Studentnum with input char* and long
//     ~UCSDStudent - to delete name
//     operator char* - get the name of the student
//     operator long == - is equal function in c++
//     Write - get the name and student number in string
//
****************************************************************************/
class UCSDStudent : public Base {
    char* name;
    long studentnum;
public:
    UCSDStudent(char* nm, long sn) :
        name(strdup(nm)), studentnum(sn) {}
    ~UCSDStudent(void) {
        /* free */
        free(name);
    }

    /* to get the name of student */
    operator char* (void) {
        return name;
    }

    /* isequal function in c++ */
    operator long(void);
    long operator == (Base& base) {
        return !strcmp(name, base);
    }

    /* write element in it as string */
    ostream& Write(ostream& stream) {
        return stream << "name: " << name
            << " Studentnum: " << studentnum;
    }

};

/*---------------------------------------------------------------------------
Function Name:                operator long
Purpose:                      to get the numeric number
Description:                  sum up the ASCII value of each character by
			      iteration
Result:                       get the ASCII value
---------------------------------------------------------------------------*/
UCSDStudent :: operator long (void) {  
    /* initialize the sum */
    long sum = 0;  
    /* use for loop to sum up */
    for (long index = 0; name[index]; index++)  
        sum += name[index];  
  
    return sum;  
}  


int main(int argc, char* const* argv) {
    SymTab ST(5);
    char buffer[128];
    char command;
    long number;
    char option;

    /* initialize debug variable */
    HashTable::Set_Debug(0);

    /* check command line options for debug display */
    while ((option = getopt(argc, argv, "x")) != EOF) {

        switch (option) {
        case 'x': HashTable::Set_Debug(1);
            break;
        }
    }

    ST.Write(cout << "Initial Symbol Table:\n");
    while (cin) {
        command = NULL;     // reset command each time in loop  
        cout << "Please enter a command:  ((i)nsert, (l)ookup, (w)rite):  ";
        cin >> command;

        switch (command) {
        case 'i':
            cout << "Please enter UCSD student name to insert:  ";
            cin >> buffer;    // formatted input  

            cout << "Please enter UCSD student number:  ";
            cin >> number;

            // create student and place in symbol table  
            if (!ST.Insert(new UCSDStudent(buffer, number))) {

                ST.Write(cout << "\nFinal Symbol Table:\n");
                exit(0);
            }
            break;

        case 'l': {
            Base* found;   // whether found or not  

            cout << "Please enter UCSD student name to lookup:  ";
            cin >> buffer;    // formatted input  

            UCSDStudent stu(buffer, 0);
            found = ST.Lookup(&stu);

            if (found)
                found->Write(cout << "Student found!!!\n") << "\n";
            else
                cout << "Student " << buffer << " not there!\n";
        }
                break;

        case 'w':
            ST.Write(cout << "The Symbol Table contains:\n");
        }
    }
    ST.Write(cout << "\nFinal Symbol Table:\n");
}

File: hw6/Driver.java
//==========================================================================
// cs12wi20cu                         Homework 6              Tianling Luo
//--------------------------------------------------------------------------
// File: Driver.java
//
// Description: In this assignment, I will write the class implementation
// for UCSDStudent which includes all UCSDStudent methods that will be 
// called by the HashTable code.
//==========================================================================

    import java.io.EOFException;  
      
//=========================================================================
// class UCSDStudent
//
// Description: The UCSDStudent class is defined and it contains the member 
// functions toString, UCSDStudent, getName, equals and hashCode to
// be called by the HashTable.java code.
//
// Data Fields:
//     name (String)  - the name of UCSD student
//     Studentnum (long)  - the student number of UCSD student
//
// Public functions:
//     toString - get the name and student number in string
//     UCSDStudent - store the name and student number in name and
//     		     Studentnum with input String and long
//     UCSDStudent - store the name and student number in name and
//     		     Studentnum with input of UCSDStudent type
//     getName - get the name of the student
//     equals - check if a student name is equal to another one
//     hashCode - get the ASCII sum of the student for a numeric value
//
//==========================================================================  
class UCSDStudent extends Base {
        private String name;    //name of student
        private long Studentnum;   //student number of student

        public String toString () {
		        //return the string with name and student number
                return "name: " + name + " Studentnum: " + Studentnum;
        }

	    public UCSDStudent (String nm, long val) {  
                //assign the name
		        name = new String (nm);  
		        //assign the Studentnum
                Studentnum = val;  
        }  
        public UCSDStudent (UCSDStudent ucsdstu) { 
	            //assign the name	
                name = new String (ucsdstu.name);
		        //assign the Studentnum
                Studentnum = ucsdstu.Studentnum;
        }  

        public String getName () {  
		        //return the name
                return name;  
        }  
        public boolean equals (Object other) {  
                if (this == other)   
                        return true;  
                  
                if (!(other instanceof UCSDStudent))   
                        return false;  
                  
                UCSDStudent otherVar = (UCSDStudent)other;  
                
		        //check if the names are the same  
                return name.equals (otherVar.getName ());  
        }  
          
        public int hashCode () {  
                int retval = 0;  
                int index = 0;  

		        //use a loop to sum up the ASCII value
                while (index != name.length ()) { 
		                //add the ASCII value	
                        retval += name.charAt (index); 
		                //increment the index	
                        index ++;  
                }  
  
                return retval;  
        }  
  
}

public class Driver {

        public static void main (String [] args) {

                /* initialize debug states */
                HashTable.debugOff();

                /* check command line options */
                for (int index = 0; index < args.length; ++index) {
                        if (args[index].equals("-x"))
                                HashTable.debugOn();
                }

                /* The real start of the code */
                SymTab symtab = new SymTab (5);
                String buffer = null;
                char command;
                long number = 0;

                System.out.print ("Initial Symbol Table:\n" + symtab);

                while (true) {
                        command = 0;    // reset command each time in loop
                        System.out.print ("Please enter a command:  "
                                + "((i)nsert, (l)ookup, (w)rite):  ");

                        try {
                        command = MyLib.getchar ();
                        MyLib.clrbuf (command); // get rid of return

                        switch (command) {
                        case 'i':
                                System.out.print (
                                "Please enter UCSD Student name to insert:  ");
                                buffer = MyLib.getline ();// formatted input

                                System.out.print (
                                        "Please enter UCSD Student number:  ");

                                number = MyLib.decin ();
                                MyLib.clrbuf (command); // get rid of return

                                // create Student and place in symbol table
                                if(!symtab.insert (
                                        new UCSDStudent (buffer, number), 1)){

                                        System.out.println("Couldn't insert " + 
                                                           "student!!!"); 
                                }
                                break;

                        case 'l': {
                                Base found;     // whether found or not

                                System.out.print (
                                "Please enter UCSD Student name to lookup:  ");

                                buffer = MyLib.getline ();// formatted input

                                UCSDStudent stu = new UCSDStudent (buffer, 0);
                                found = symtab.lookup (stu);
                                
                                if (found != null) {
                                        System.out.println ("Student found!!!");
                                        System.out.println (found);
                                }
                                else
                                        System.out.println ("Student " + buffer
                                                + " not there!");
                                }
                                break;

                        case 'w':
                                System.out.print (
                                    "The Symbol Table contains:\n" + symtab);
                        }
                        }
                        catch (EOFException eof) {
                                break;
                        }
                }

                System.out.print ("\nFinal Symbol Table:\n" + symtab);
        }
}
